File: frontend/src/game/upgrades/stat_modifiers/knockback_2.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const Knockback2Def: UpgradeDef = {
  id: "knockback_2",
  name: "Knockback Boost",
  description: "+20% knockback",
  rarity: RarityID.Uncommon,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 6,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Knockback,
  value: 0.2,
  isMultiplier: true,
  stackable: true,
  maxStacks: 3,
}

export class Knockback2 extends Upgrade {}

File: backend/app/core/upgrades/stat_modifiers/knockback_2.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Knockback2(UpgradeImplementation):
    id = 'knockback_2'
    name = 'Knockback Boost'
    description = '+20% knockback'
    rarity = 'uncommon'
    type = 'stat_modifier'
    cost = 6

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.2
        appliers.modifiers.add_multiplier('attack', 'knockback', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
